File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/filter/telegram/UserArrayFilterParser.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.telegram;

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.BooleanFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.UserArrayFilter;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.impl.UnionExtractValidatorList;
import org.springframework.stereotype.Component;
import io.github.oleksivio.telegram.bot.api.model.objects.std.User;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Component
public class UserArrayFilterParser implements FilterParser<UserArrayFilter, List<User>> {
    @Override
    public Validator<List<User>> createChecker(UserArrayFilter annotation, Finder finder) {

        UnionExtractValidatorList<User> unionExtractValidatorList = new UnionExtractValidatorList<>();

        Arrays.stream(annotation.validator())
                .map(validatorName -> finder.find(validatorName,User.class))
                .forEach(validator -> unionExtractValidatorList.add(Optional::of, validator));

        StringFilter firstName = annotation.firstName();
        if (firstName.status().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getFirstName()), finder.find(firstName));
        }
        StringFilter lastName = annotation.lastName();
        if (lastName.status().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getLastName()), finder.find(lastName));
        }
        StringFilter username = annotation.username();
        if (username.status().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getUsername()), finder.find(username));
        }
        StringFilter languageCode = annotation.languageCode();
        if (languageCode.status().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getLanguageCode()), finder.find(languageCode));
        }
        BooleanFilter bot = annotation.bot();
        if (bot.status().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getBot()), finder.find(bot));
        }

        return unionExtractValidatorList;
    }

    @Override
    public Class<UserArrayFilter> getAnnotationClass() {
        return UserArrayFilter.class;
    }
}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/annotations/parser/filter/telegram/UserArrayFilterParser.kt
package io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.tl.kbot.core.annotations.filter.composite.UserArrayFilter
import io.github.oleksivio.tl.kbot.server.api.objects.std.User
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.impl.UnionExtractValidatorList
 

  
class UserArrayFilterParser :
    FilterParser<UserArrayFilter, List<User>> {

    override val parserAnnotationClass = UserArrayFilter::class

    override fun createChecker(annotation: UserArrayFilter, finder: Finder): Validator<List<User>> {

        val unionExtractValidatorList =
            UnionExtractValidatorList<User>()

        annotation.validator
                .map { validatorName -> finder.find(validatorName, User::class) }
                .forEach { validator -> unionExtractValidatorList.add({ it }, validator) }

        val firstName = annotation.firstName
        if (firstName.status.isActive) {
            unionExtractValidatorList.add({ it.firstName }, finder.find(firstName))
        }
        val lastName = annotation.lastName
        if (lastName.status.isActive) {
            unionExtractValidatorList.add({ it.lastName }, finder.find(lastName))
        }
        val username = annotation.username
        if (username.status.isActive) {
            unionExtractValidatorList.add({ it.username }, finder.find(username))
        }
        val languageCode = annotation.languageCode
        if (languageCode.status.isActive) {
            unionExtractValidatorList.add({ it.languageCode }, finder.find(languageCode))
        }
        val bot = annotation.bot
        if (bot.status.isActive) {
            unionExtractValidatorList.add({ it.bot }, finder.find(bot))
        }

        return unionExtractValidatorList
    }
}
